File: src/parseActivity.ts
import {
  type ConversationKey,
  type InboundEvent,
  type InboundMessage,
  type InboundReaction,
  type MessageId,
  brandConversationKey,
  brandMessageId,
  isRecord,
  readString,
} from "./types.js";

export type ParseOk = { kind: "ok"; event: InboundEvent };
export type ParseErr = { kind: "error"; message: string };
export type ParseResult = ParseOk | ParseErr;

function conversationKeyOf(
  conversationId: string,
  threadHint: string | undefined,
): ConversationKey {
  if (threadHint !== undefined && threadHint.length > 0 && !conversationId.includes(";messageid=")) {
    return brandConversationKey(`${conversationId};thread=${threadHint}`);
  }
  return brandConversationKey(conversationId);
}

function readConversation(value: unknown): { id: string; conversationType: string } | undefined {
  if (!isRecord(value)) {
    return undefined;
  }
  const id = readString(value.id);
  if (id === undefined || id.length === 0) {
    return undefined;
  }
  const conversationType = readString(value.conversationType) ?? "personal";
  return { id, conversationType };
}

function readFromId(value: unknown): string {
  if (!isRecord(value)) {
    return "unknown";
  }
  return readString(value.id) ?? "unknown";
}

function readReplyToId(value: unknown): MessageId | undefined {
  const raw = readString(value);
  if (raw === undefined || raw.length === 0) {
    return undefined;
  }
  return brandMessageId(raw);
}

function readThreadHint(activity: Record<string, unknown>): string | undefined {
  const conversation = activity.conversation;
  if (isRecord(conversation)) {
    const threadId = readString(conversation.threadId);
    if (threadId !== undefined && threadId.length > 0) {
      return threadId;
    }
  }
  const channelData = activity.channelData;
  if (isRecord(channelData)) {
    const thread = channelData.thread;
    if (isRecord(thread)) {
      const id = readString(thread.id);
      if (id !== undefined && id.length > 0) {
        return id;
      }
    }
  }
  return undefined;
}

function readTeamChannel(activity: Record<string, unknown>): { teamId?: string; channelId?: string } {
  const channelData = activity.channelData;
  if (!isRecord(channelData)) {
    return {};
  }
  const out: { teamId?: string; channelId?: string } = {};
  const team = channelData.team;
  if (isRecord(team)) {
    const teamId = readString(team.id);
    if (teamId !== undefined && teamId.length > 0) {
      out.teamId = teamId;
    }
  }
  const channel = channelData.channel;
  if (isRecord(channel)) {
    const channelId = readString(channel.id);
    if (channelId !== undefined && channelId.length > 0) {
      out.channelId = channelId;
    }
  }
  return out;
}

function parseMessage(activity: Record<string, unknown>): ParseResult {
  const conversation = readConversation(activity.conversation);
  if (conversation === undefined) {
    return { kind: "error", message: "message activity missing conversation.id" };
  }
  const id = readString(activity.id);
  if (id === undefined || id.length === 0) {
    return { kind: "error", message: "message activity missing id" };
  }
  const text = readString(activity.text) ?? "";
  const serviceUrl = readString(activity.serviceUrl) ?? "";
  const replyToId = readReplyToId(activity.replyToId);
  const ids = readTeamChannel(activity);
  const event: InboundMessage = {
    kind: "message",
    messageId: brandMessageId(id),
    text,
    conversationKey: conversationKeyOf(conversation.id, readThreadHint(activity)),
    conversationId: conversation.id,
    serviceUrl,
    fromId: readFromId(activity.from),
    conversationType: conversation.conversationType,
  };
  if (replyToId !== undefined) {
    event.replyToId = replyToId;
  }
  if (ids.teamId !== undefined) {
    event.teamId = ids.teamId;
  }
  if (ids.channelId !== undefined) {
    event.channelId = ids.channelId;
  }
  return { kind: "ok", event };
}

function reactionsFrom(value: unknown): string[] {
  if (!Array.isArray(value)) {
    return [];
  }
  const emojis: string[] = [];
  for (const item of value) {
    if (!isRecord(item)) {
      continue;
    }
    const emoji = readString(item.type);
    if (emoji !== undefined && emoji.length > 0) {
      emojis.push(emoji);
    }
  }
  return emojis;
}

function buildReactionEvents(body: Record<string, unknown>): InboundReaction[] {
  const conversation = readConversation(body.conversation);
  if (conversation === undefined) {
    return [];
  }
  const target = readReplyToId(body.replyToId);
  if (target === undefined) {
    return [];
  }
  const fromId = readFromId(body.from);
  const serviceUrl = readString(body.serviceUrl) ?? "";
  const key = conversationKeyOf(conversation.id, readThreadHint(body));
  const out: InboundReaction[] = [];
  for (const emoji of reactionsFrom(body.reactionsAdded)) {
    out.push({
      kind: "reaction",
      messageId: target,
      emoji,
      action: "add",
      fromId,
      conversationKey: key,
      conversationId: conversation.id,
      serviceUrl,
    });
  }
  for (const emoji of reactionsFrom(body.reactionsRemoved)) {
    out.push({
      kind: "reaction",
      messageId: target,
      emoji,
      action: "remove",
      fromId,
      conversationKey: key,
      conversationId: conversation.id,
      serviceUrl,
    });
  }
  return out;
}

function parseReaction(activity: Record<string, unknown>): ParseResult {
  if (readConversation(activity.conversation) === undefined) {
    return { kind: "error", message: "messageReaction missing conversation.id" };
  }
  if (readReplyToId(activity.replyToId) === undefined) {
    return { kind: "error", message: "messageReaction missing replyToId (target message)" };
  }
  const events = buildReactionEvents(activity);
  const first = events[0];
  if (first === undefined) {
    return { kind: "error", message: "messageReaction has no reactionsAdded or reactionsRemoved" };
  }
  return { kind: "ok", event: first };
}

export function parseReactionAll(body: unknown): InboundReaction[] {
  if (!isRecord(body)) {
    return [];
  }
  if (readString(body.type) !== "messageReaction") {
    return [];
  }
  return buildReactionEvents(body);
}

const IGNORED_TYPES = new Set([
  "conversationUpdate",
  "installationUpdate",
  "invoke",
  "typing",
  "event",
]);

export function parseActivity(body: unknown): ParseResult {
  if (!isRecord(body)) {
    return { kind: "error", message: "activity is not an object" };
  }
  const type = readString(body.type);
  if (type === undefined) {
    return { kind: "error", message: "activity missing type" };
  }
  if (type === "message") {
    return parseMessage(body);
  }
  if (type === "messageReaction") {
    return parseReaction(body);
  }
  if (IGNORED_TYPES.has(type)) {
    return {
      kind: "ok",
      event: { kind: "ignored", type, reason: `activity type ${type} does not start a turn` },
    };
  }
  return {
    kind: "ok",
    event: { kind: "ignored", type, reason: `unhandled activity type ${type}` },
  };
}

File: python/teams_harness/parse_activity.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from .types import (
    ConversationKey,
    InboundEvent,
    InboundIgnored,
    InboundMessage,
    InboundReaction,
    MessageId,
    brand_conversation_key,
    brand_message_id,
    is_record,
    read_string,
)

IGNORED_TYPES = {
    "conversationUpdate",
    "installationUpdate",
    "invoke",
    "typing",
    "event",
}


@dataclass(slots=True, frozen=True)
class ParseOk:
    kind: str
    event: InboundEvent


@dataclass(slots=True, frozen=True)
class ParseErr:
    kind: str
    message: str


ParseResult = ParseOk | ParseErr


def conversation_key_of(conversation_id: str, thread_hint: str | None) -> ConversationKey:
    if thread_hint and ";messageid=" not in conversation_id:
        return brand_conversation_key(f"{conversation_id};thread={thread_hint}")
    return brand_conversation_key(conversation_id)


def _read_conversation(value: object) -> tuple[str, str] | None:
    if not is_record(value):
        return None
    conv_id = read_string(value.get("id"))
    if not conv_id:
        return None
    conversation_type = read_string(value.get("conversationType")) or "personal"
    return conv_id, conversation_type


def _read_from_id(value: object) -> str:
    if not is_record(value):
        return "unknown"
    return read_string(value.get("id")) or "unknown"


def _read_reply_to_id(value: object) -> MessageId | None:
    raw = read_string(value)
    if not raw:
        return None
    return brand_message_id(raw)


def _read_thread_hint(activity: dict[str, Any]) -> str | None:
    conversation = activity.get("conversation")
    if is_record(conversation):
        thread_id = read_string(conversation.get("threadId"))
        if thread_id:
            return thread_id
    channel_data = activity.get("channelData")
    if is_record(channel_data):
        thread = channel_data.get("thread")
        if is_record(thread):
            thread_id = read_string(thread.get("id"))
            if thread_id:
                return thread_id
    return None


def _read_team_channel(activity: dict[str, Any]) -> tuple[str | None, str | None]:
    channel_data = activity.get("channelData")
    if not is_record(channel_data):
        return None, None
    team_id = None
    channel_id = None
    team = channel_data.get("team")
    if is_record(team):
        team_id = read_string(team.get("id")) or None
    channel = channel_data.get("channel")
    if is_record(channel):
        channel_id = read_string(channel.get("id")) or None
    return team_id, channel_id


def _parse_message(activity: dict[str, Any]) -> ParseResult:
    conversation = _read_conversation(activity.get("conversation"))
    if conversation is None:
        return ParseErr(kind="error", message="message activity missing conversation.id")
    conv_id, conversation_type = conversation
    msg_id = read_string(activity.get("id"))
    if not msg_id:
        return ParseErr(kind="error", message="message activity missing id")
    text = read_string(activity.get("text")) or ""
    service_url = read_string(activity.get("serviceUrl")) or ""
    reply_to_id = _read_reply_to_id(activity.get("replyToId"))
    team_id, channel_id = _read_team_channel(activity)
    event = InboundMessage(
        kind="message",
        message_id=brand_message_id(msg_id),
        text=text,
        conversation_key=conversation_key_of(conv_id, _read_thread_hint(activity)),
        conversation_id=conv_id,
        service_url=service_url,
        from_id=_read_from_id(activity.get("from")),
        conversation_type=conversation_type,
        reply_to_id=reply_to_id,
        team_id=team_id,
        channel_id=channel_id,
    )
    return ParseOk(kind="ok", event=event)


def _reactions_from(value: object) -> list[str]:
    if not isinstance(value, list):
        return []
    emojis: list[str] = []
    for item in value:
        if is_record(item):
            emoji = read_string(item.get("type"))
            if emoji:
                emojis.append(emoji)
    return emojis


def _build_reaction_events(body: dict[str, Any]) -> list[InboundReaction]:
    conversation = _read_conversation(body.get("conversation"))
    if conversation is None:
        return []
    conv_id, _conversation_type = conversation
    target = _read_reply_to_id(body.get("replyToId"))
    if target is None:
        return []
    from_id = _read_from_id(body.get("from"))
    service_url = read_string(body.get("serviceUrl")) or ""
    key = conversation_key_of(conv_id, _read_thread_hint(body))
    out: list[InboundReaction] = []
    for emoji in _reactions_from(body.get("reactionsAdded")):
        out.append(
            InboundReaction(
                kind="reaction",
                message_id=target,
                emoji=emoji,
                action="add",
                from_id=from_id,
                conversation_key=key,
                conversation_id=conv_id,
                service_url=service_url,
            )
        )
    for emoji in _reactions_from(body.get("reactionsRemoved")):
        out.append(
            InboundReaction(
                kind="reaction",
                message_id=target,
                emoji=emoji,
                action="remove",
                from_id=from_id,
                conversation_key=key,
                conversation_id=conv_id,
                service_url=service_url,
            )
        )
    return out


def parse_reaction_all(body: object) -> list[InboundReaction]:
    if not is_record(body):
        return []
    if read_string(body.get("type")) != "messageReaction":
        return []
    return _build_reaction_events(body)


def parse_activity(body: object) -> ParseResult:
    if not is_record(body):
        return ParseErr(kind="error", message="activity is not an object")
    activity_type = read_string(body.get("type"))
    if activity_type is None:
        return ParseErr(kind="error", message="activity missing type")
    if activity_type == "message":
        return _parse_message(body)
    if activity_type == "messageReaction":
        if _read_conversation(body.get("conversation")) is None:
            return ParseErr(kind="error", message="messageReaction missing conversation.id")
        if _read_reply_to_id(body.get("replyToId")) is None:
            return ParseErr(kind="error", message="messageReaction missing replyToId (target message)")
        events = _build_reaction_events(body)
        if not events:
            return ParseErr(kind="error", message="messageReaction has no reactionsAdded or reactionsRemoved")
        return ParseOk(kind="ok", event=events[0])
    if activity_type in IGNORED_TYPES:
        return ParseOk(
            kind="ok",
            event=InboundIgnored(kind="ignored", type=activity_type, reason=f"activity type {activity_type} does not start a turn"),
        )
    return ParseOk(
        kind="ok",
        event=InboundIgnored(kind="ignored", type=activity_type, reason=f"unhandled activity type {activity_type}"),
    )
